### PR TITLE
Respect cluster alias in `_index` aggs and queries

### LINKED ISSUE #25606: Cross cluster aggregations don't include cluster alias in _index
**Elasticsearch version**: master
**Plugins installed**: none
**JVM version**: 1.8.0_121
**OS version**: macOS

**Description of the problem including expected versus actual behavior**:
When running a search on a remote cluster, remote documents include the cluster alias in the `_index` field so that the document can be properly addressed. This behavior does not apply to aggregations. When doing a `terms` agg on `_index` for instance, the terms do not include the cluster alias. The `top_hits` agg also doesn't enhance `_index` with the cluster alias.

**Steps to reproduce**:
 1. Start a node that remotes to itself

     ```sh
     ./bin/elasticsearch -E search.remote.local.seeds=localhost:9300
     ```

 2. Index a document

      ```sh
      curl -XPOST "http://localhost:9200/index/doc/id" -H 'Content-Type: application/json' -d'
      {
        "foo": "bar"
      }'
      ```

 3. Execute a search for the document via local and remote index names, which results in two unique documents (correct behavior) but one index term in the aggs and what appears to be two identical `top_hits`

      ```sh
    curl -XPOST "http://localhost:9200/index,local:index/_search" -H 'Content-Type: application/json' -d'
      {
        "aggs": {
          "indices": {
            "terms": {
              "field": "_index"
            },
            "aggs": {
              "hits": {
                "top_hits": {
                  "size": 10
                }
              }
            }
          }
        }
      }'
      ```

      results:
      ```json
      {
        "took": 7,
        "timed_out": false,
        "_shards": {
          "total": 10,
          "successful": 10,
          "failed": 0
        },
        "hits": {
          "total": 2,
          "max_score": 1,
          "hits": [
            {
              "_index": "index",
              "_type": "doc",
              "_id": "id",
              "_score": 1,
              "_source": {
                "foo": "bar"
              }
            },
            {
              "_index": "local:index",
              "_type": "doc",
              "_id": "id",
              "_score": 1,
              "_source": {
                "foo": "bar"
              }
            }
          ]
        },
        "aggregations": {
          "indices": {
            "doc_count_error_upper_bound": 0,
            "sum_other_doc_count": 0,
            "buckets": [
              {
                "key": "index",
                "doc_count": 2,
                "hits": {
                  "hits": {
                    "total": 2,
                    "max_score": 1,
                    "hits": [
                      {
                        "_index": "index",
                        "_type": "doc",
                        "_id": "id",
                        "_score": 1,
                        "_source": {
                          "foo": "bar"
                        }
                      },
                      {
                        "_index": "index",
                        "_type": "doc",
                        "_id": "id",
                        "_score": 1,
                        "_source": {
                          "foo": "bar"
                        }
                      }
                    ]
                  }
                }
              }
            ]
          }
        }
      }
      ```

### PR DESCRIPTION
**Pull request by s1monw**
Today when we aggregate on the `_index` field the cross cluster search
alias is not taken into account. Neither is it respected when we search
on the field. This change adds support for cluster alias when the cluster
alias is present on the `_index` field.

Closes #25606

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 1]
```diff
@@ -467,17 +467,14 @@ public IndexSettings getIndexSettings() {
...
+        return new QueryShardContext(shardId, indexSettings, indexCache.bitsetFilterCache(), indexFieldData::getForField, mapperService(),
+            similarityService(), scriptService, xContentRegistry, client, indexReader, nowInMillis, clusterAlias);
     }
 
-    /**
-     * The {@link ThreadPool} to use for this index.
-     */
+        /**
+         * The {@link ThreadPool} to use for this index.
+         */
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: indentation was correct before

**Review by jpountz — APPROVED** [review]
It is a bit of a hack but it looks sustainable. I'm wondering that we might be able to simplify things a bit by making the MappedFieldType responsible for producing an IndexFieldData instance directly instead of having an intermediate builder, and passing a QueryShardContext to this method like we do for factory methods of queries.

**Comment by s1monw** [comment]
> It is a bit of a hack but it looks sustainable. I'm wondering that we might be able to simplify things a bit by making the MappedFieldType responsible for producing an IndexFieldData instance directly instead of having an intermediate builder, and passing a QueryShardContext to this method like we do for factory methods of queries.

this might be a good way to make this simpler. I will look into it. thanks!

**Comment by s1monw** [comment]
> It is a bit of a hack but it looks sustainable. I'm wondering that we might be able to simplify things a bit by making the MappedFieldType responsible for producing an IndexFieldData instance directly instead of having an intermediate builder, and passing a QueryShardContext to this method like we do for factory methods of queries.

@jpountz @jimczi are you ok if we do this as a followup? I think the the place we have hack in is pretty contained and we can move on with this as is?

**Review by jimczi — APPROVED** [review]
LGTM too

**Comment by jpountz** [comment]
Sure, let's do this as a follow-up.

**Pull request merged**